File: CommonModule/src/main/java/com/qihuan/commonmodule/base/BaseApp.java
package com.qihuan.commonmodule.base;

import android.app.Application;

import com.qihuan.commonmodule.collection.bean.MyObjectBox;
import com.qihuan.commonmodule.net.ApiClient;
import com.qihuan.commonmodule.utils.AppUtils;
import com.qihuan.commonmodule.utils.ToastUtils;

import io.objectbox.BoxStore;

/**
 * BaseApp
 *
 * @author Qi
 */
public class BaseApp extends Application {

    private static BaseApp app;
    private BoxStore boxStore;

    @Override
    public void onCreate() {
        super.onCreate();
        app = this;
        init();
    }

    public static BaseApp getInstance() {
        return app;
    }

    private void init() {
        ApiClient.init(this);
        AppUtils.init(this);
        ToastUtils.init(this);

        boxStore = MyObjectBox.builder()
            .androidContext(this)
            .build();
    }

    public BoxStore getBoxStore() {
        return boxStore;
    }
}

File: module_common/src/main/java/com/qihuan/commonmodule/base/BaseApp.kt
package com.qihuan.commonmodule.base

import android.app.Application
import com.alibaba.android.arouter.launcher.ARouter
import com.github.moduth.blockcanary.BlockCanary
import com.qihuan.commonmodule.BuildConfig
import com.qihuan.commonmodule.net.ApiManager
import com.scwang.smartrefresh.header.DeliveryHeader
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.footer.ClassicsFooter
import com.squareup.leakcanary.LeakCanary
import kotlin.properties.Delegates


/**
 * BaseApp
 *
 * @author Qi
 */
open class BaseApp : Application() {

    override fun onCreate() {
        super.onCreate()
        instance = this
        init()
    }

    private fun init() {
        ApiManager.init(this)
        if (BuildConfig.DEBUG) {
            ARouter.openLog()
            ARouter.openDebug()
            ARouter.printStackTrace()
        }
        ARouter.init(this)
        if (BuildConfig.DEBUG) {
            // LeakCanary
            if (LeakCanary.isInAnalyzerProcess(this)) {
                return
            }
            LeakCanary.install(this)
            // BlockCanary
            BlockCanary.install(this, AppBlockCanaryContext()).start()
        }
    }

    companion object {
        var instance: BaseApp by Delegates.notNull()

        init {
            // 设置全局的Header构建器
            SmartRefreshLayout.setDefaultRefreshHeaderCreator { context, _ ->
                // 指定为经典Header，默认是 贝塞尔雷达Header
                DeliveryHeader(context)
            }
            // 设置全局的Footer构建器
            SmartRefreshLayout.setDefaultRefreshFooterCreator { context, _ ->
                // 指定为经典Footer，默认是 BallPulseFooter
                ClassicsFooter(context).setDrawableSize(20f)
            }
        }
    }
}
